### ISSUE TITLE
Cannot add feeds to Django Community Blog

### ISSUE THREAD

**Issue by ndunn219** [django/djangoproject.com#1137]
There seems to be an issue adding feeds to the Django Community Blog. My feed (https://www.webucator.com/articles/django/feed/) was approved in October by @adamchainz, but hasn't been added. Adam thought there might be an issue with https://superfeedr.com/, which he said was sending out some obscure certificate errors.

**Comment by carltongibson** [comment]
From time to time we see an error here, but I'm at a loss as to how to debug. 

I'm wondering if superfeedr.com is a reliable dependency here? 🤔

**Comment by adamchainz** [comment]
I tried to approve another feed yesterday and it also failed - is that the error you saw? Could you share a stack trace?

**Comment by carltongibson** [comment]
```
SubscriptionError: ('Error during request to hub https://push.superfeedr.com/ for topic https://carrick.eu/feeds/django.atom.xml: We could not verify your callback Error: certificate has expired', <Subscription: Subscription object (1010)>, <Response [422]>)
  File "django-admin.py", line 21, in <module>
    management.execute_from_command_line()
  File "django/core/management/__init__.py", line 419, in execute_from_command_line
    utility.execute()
  File "django/core/management/__init__.py", line 413, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "django/core/management/base.py", line 354, in run_from_argv
    self.execute(*args, **cmd_options)
  File "django/core/management/base.py", line 398, in execute
    output = self.handle(*args, **options)
  File "aggregator/management/commands/update_subscriptions.py", line 33, in handle
    subscription.unsubscribe()
  File "django_push/subscriber/models.py", line 94, in unsubscribe
    return self.send_request(mode='unsubscribe')
  File "django_push/subscriber/models.py", line 135, in send_request
    raise SubscriptionError(
```

> We could not verify your callback Error: certificate has expired

Now is that them saying our certificate has expired... 🤔

**Comment by carltongibson** [comment]
Error is being raised in Django-push here: 

https://github.com/brutasse/django-push/blob/ee83ff553aa4bdf2ff969f5443971450cdd25eb5/django_push/subscriber/models.py#L136

Hitting the callback URL there **doesn't raise** the SSL error — our certificate is fine AFAICS. 
🤔

**Comment by carltongibson** [comment]
(We don't have a lot of feeds. Fetching them each once a day ourselves might be better than a flaky dependency...)

**Comment by adamchainz** [comment]
Superfeedr looks like a useful service since it can normalize between the different feed types. But perhaps there is something up with the requests django-push is making, or perhaps our credentials have expired? Can you double check that?

**Comment by carltongibson** [comment]
@adamchainz I think this is Superfeedr reporting an error when trying to send a request to our callback URL. (Not us making a request to them.)

AFAICS they’re reporting a certificate issue, where testing the same thing manually none exists.

**Comment by adamchainz** [comment]
Yeah that is what it is, I just wanted to be sure. I found their support is on GitHub and that there's [this issue](https://github.com/superfeedr/documentation/issues/155) open since November covering the same problem. No reply from them, so I guess they aren't maintained. To be sure, I've emailed the support address, but I would guess it's time to try migrate off indeed.

**Comment by carltongibson** [comment]
Might be good for the DjangoCon sprints if we don't make it before.

**Comment by carltongibson** [comment]
Again, this isn't stale. Rather we need to work out how to replace the super feeder service 

(A cron job with feedparser may be enough here 🤔)

**Comment by julien51** [comment]
Would you be able to help me with the certificate issues?

**Comment by salihslx** [comment]
Hai

**Comment by thibaudcolas** [comment]
Just checking if anyone who’s been looking into this so far can provide an update on where this is at, and what they’d suggest we do?

I’ve always found it bizarre how Django’s community blog aggregator only seemed to showcase the work of very few people. I thought it was a strange decision / legacy thing, rather than just software being broken, until talking to @adamchainz a few days ago.

**Comment by adamchainz** [comment]
I don't believe anyone has looked at it. This issue is all the progress I know of.

**Comment by carltongibson** [comment]
Hey @thibaudcolas — Yes... I've got a folder somewhere with a few `feedparser` noodles in it, but nothing more than that I'm afraid. 🤹

This would be an awesome project for someone to pick up — it's a real pain deficit not having the aggregator working.

**Comment by thibaudcolas** [comment]
@carltongibson are you ok to share those even if it’s just notes?

From my perspective I see four options to getting fixed in 2024:

1. Have the proposed [website working group](https://github.com/django/dsf-working-groups/issues/2) take ownership of this and find a volunteer maintainer as part of that group to take it on.
2. Retire the feed(s) and replace them with something that’s lower maintenance, depending on the findings of the [Django website UX project](https://forum.djangoproject.com/t/at-what-point-is-the-work-on-the-ux-of-the-django-website/25981) led by @pauloxnet
3. Propose this as a Google Summer of Code project idea in 2024. Or as part of a wider project of "Django website improvements"
4. Find someone interested in helping with the website as part of [Djangonaut Space](https://djangonaut.space/)

What do you think of those options? Any other ones we should consider? Any specific person you’d think isn’t in this thread and might be able to help?

**Comment by carltongibson** [comment]
@thibaudcolas, sure. 

Feedparser is the standard Python tool here: https://feedparser.readthedocs.io/en/latest/index.html

Basic usage is fairly straight-forward: 

```pycon
>>> import feedparser 
>>> d = feedparser.parse('https://noumenal.es/posts/feed/')
>>> d['feed']['title']
"Carlton's latest posts."
>>> d.entries[0].title
'The longest year'
```

The relevant models are in the `aggregator` app. `Feed` and `FeedItem`.

https://github.com/django/djangoproject.com/blob/main/aggregator/models.py

The task would be to loop the approved `Feed` objects, fetch the data, and create the appropriate `FeedItem` objects for each one. I guess: 

* A function that does it for one. 
* Then looping that in a management command. 

Feedparser supports [Etag and Last-Modified](https://feedparser.readthedocs.io/en/latest/http-etag.html#etag-and-last-modified-headers), so (maybe 🤔) accepting a `--since` flag would be useful. We could then track `last-run` (like in a txt file would do — it doesn't have to be clever). 

I would then run that with a systemd timer unit every few hours, but there are a billion options, so I'd defer to whatever Ops fancies. 

That seems to me like it should be do-able to us. If the running it were an issue — it's a small number of feeds fetched periodically... — I'd add an API endpoint to create the `FeedItems` and I'd happily run it myself (as I imagine would many others). 

--- 

Of the options you mention, 4 stands out to me as an obvious one. Anyone up for it? 😜 I would happily input on a PR (or whatever is needed). 

I don't think it's a **massive** project. It does need someone to clear the decks for it, to give it time. No doubt, a prototype would be simple enough, but then there'll be errors that come up, and need fixing. (I'd suggest adding a "We can't fetch this" flag to `Feed`...) 


That's more or less all I've got. 🎄

**Comment by adamchainz** [comment]
Thanks for all the information Carlton. And yes, this would be great for someone to pick up, if  Djangonaut Space can help with that it would be awesome.

> (like in a txt file would do — it doesn't have to be clever).

I'd suggest a model instead. The change can be committed in a transaction along with the fetched items. And also the tracking data wouldn't be lost if we rebuild the server. This could be a [single instance model](https://adamj.eu/tech/2021/02/04/django-check-constraints-limit-model-single-instance/).

**Comment by carltongibson** [comment]
Ok, yes that's a neat idea. 💡 (It's just a fetching optimisation... you still have to check the feed items don't already exist, because feed dates aren't always reliable, and servers don't always support Last-Modified &co, but yes 👍)

**Comment by Pradhvan** [comment]
Based on the suggestions mentioned above. I would like to work on it.  😊

👀 Let me go through the entire thread and comment out a plan before I push any code for it.

**Comment by carltongibson** [comment]
@Pradhvan sounds great! 

The management command... Yes, literally it needs to be run periodically, fetching  feeds for updates. (What you described is right, so is there a clarification you need?)

You could imagine doing that by hand, but we'll schedule it using cron or a systemd timer unit or such. Having a command gives the hook to connect that. 

Make sense?

**Comment by Pradhvan** [comment]
> The management command... Yes, literally it needs to be run periodically, fetching feeds for updates. 

oh okay, got that. A hook to perform action in a scheduled manner. 

> (What you described is right, so is there a clarification you need?)
yup, just wanted to confirm that I am going in the right direction. 😄 

I think I have much clear picture now, let me work the plan discussed on my fork and come back with draft PR/demo or something like that. 

Thank @carltongibson for a quick response. 🤗

**Comment by nanuxbe** [comment]
@Pradhvan are you still looking into this?
If you're not, I will try to find time to give it a go in the coming week.

**Comment by jefftriplett** [comment]
I closed #1299 because feeds appear to be fixed and working again. Noting it here in case this issue is resolved too.

**Comment by thibaudcolas** [comment]
@knyghty will test this with his blog and report back ⭐️

**Comment by adamchainz** [comment]
I haven't received @knyghty 's feed, but I have seen a few feed submissions recently, and I approved most of them. I think the system is working fine now... (?)

**Comment by adamzap** [comment]
@knyghty Were you able to test this?

**Comment by adamchainz** [comment]
Looks like my April 29 post was mistaken, as when I check the admin, I actually did approve @knyghty 's feed on April 19! My bad there.

Since everything seems to be working, can we close this thread?

**Comment by adamzap** [comment]
I think so, thanks! Please reopen it if there is still an issue.

**Comment by knyghty** [comment]
Hmm I posted something to test yesterday (finally) and it still doesn't show up. So this is either an issue or I did something wrong.
